### ISSUE TITLE
Refactor QueryHandler Initialization

### ISSUE THREAD

**Issue by blentz** [project-koku/koku#960]
This issue was brought into focus while working on #938. 

The QueryHandler object tree has a really brittle, complex initialization sequence. This makes working on changes in this area more complex and difficult than it probably needs to be.

Of particular note are the steps that depend upon looking up the group_by parameter to make decisions.

This issue is to redesign and refactor QueryHandler initialization.

Proposed change:

1. Create a new object to hold data values passed into the QueryHandler from the Views. Let's call it "QueryData" for now. This object will be responsible for taking values from the `_generic_report()` view function, processing those values into the form required by the QueryHandler, and providing a set of getters or properties that the QueryHandler can use to access data coming in from the views.

This "QueryData" object is expected to replace the `url_data`, `query_parameters`, and `kwargs` parameters currently required by the QueryHandler. Instead of passing multiple individual data points, we'll pass a single QueryData object into the QueryHandler. 

The information contained within the QueryData should be "ready to go" and require no additional processing by the QueryHandler. (i.e. QueryHandler treats QueryData as read-only.)

2. Untangle the initialization flow. The basic flow for the whole QueryHandler should look something like this:

a. Receive parsed view data from the constructor parameters.
b. Look up model information required to compile the report out of the ProviderMap
c. Compile the requested query (i.e. `execute_query()`)

So, as part of refactoring the initialization portion of that workflow, we need to distinguish between View Data, Model References, and any metadata (e.g. class attributes) that are inherent to the QueryHandler objects themselves.

**Comment by adberglund** [comment]
Since there is refactor involved here let's discuss as a team before starting work on this and agree on the proposal before beginning work. 

I put this card in the Design column so it would not get lost in the backlog or new issues (both columns are bloated). I don't anticipate getting to this until after Azure, and wanted to keep it visible on the board so we didn't lose track of it. Just trying something new.

**Comment by blentz** [comment]
We discussed this and agreed to do a small/quick (1-2 hours) POC of the idea to help show the code impact. Once a POC is ready, we'll meet again to discuss sizing and timing on a full implementation.

**Comment by blentz** [comment]
I've pushed a branch to show off some code around the QueryHandler Initialization discussion:
https://github.com/project-koku/koku/compare/960_queryhandler

This isn't fully functional quite yet, but it should at least show off the idea enough for us to discuss a go/no-go decision and stuff like timing/priority.

The main highlights are:
- The flow from View into the QueryHandler is quite a bit simpler. The QueryHandler's initialization sequence is also significantly less order-dependent.

- I'm able to do away with one of the mapping objects, partly thanks to Bihan's work moving to class-based views. That simplifies things quite a bit. (I could extract this piece into a smaller PR; it doesn't have any dependencies on the rest of the changes).

- I've moved most of the logic around munging the Request object's data into its own object - QueryParameters. This enables me to move and/or delete some of the helper functions in the QueryHandler that were doing additional parameter processing. It looks like it's improving the separation of concerns and DRYs up the code a fair bit.
